### ISSUE TITLE
Feature request: device function arguments in OpenCL

### ISSUE THREAD

**Issue by tcew** [libocca/occa#166]
Device functions need to inherit memory space and `const` qualifiers from the calling kernel.

Example errors:

```
OpenCL compiling ellipticAxIpdgHex3D from [/root/.occa/cache/f4e14a1457675b95/source.cpp] with compiler flags [-cl-opt-disable]
/tmp/AMD_3002_379/t_3002_381.cl:144:50: error: passing 'const **__global** double *' to parameter of type 'const **double** *' changes address space of pointer
        surfaceTerms(emap, sk0, 0, 0, i, j, tau, sgeo, vmapM, vmapP, EToB, gradq, s_dqdx, s_dqdy, s_dqdz, s_Aq);
                                                 ^~~~
/tmp/AMD_3002_379/t_3002_381.cl:28:33: note: passing argument to parameter 'sgeo' here
                  const double *sgeo,
                                ^
/tmp/AMD_3002_379/t_3002_381.cl:144:56: error: passing 'const __global int *' to parameter of type 'const int *' changes address space of pointer
        surfaceTerms(emap, sk0, 0, 0, i, j, tau, sgeo, vmapM, vmapP, EToB, gradq, s_dqdx, s_dqdy, s_dqdz, s_Aq);
                                                       ^~~~~
/tmp/AMD_3002_379/t_3002_381.cl:29:30: note: passing argument to parameter 'vmapM' here
                  const int *vmapM,
                             ^
/tmp/AMD_3002_379/t_3002_381.cl:144:63: error: passing 'const __global int *' to parameter of type 'const int *' changes address space of pointer
        surfaceTerms(emap, sk0, 0, 0, i, j, tau, sgeo, vmapM, vmapP, EToB, gradq, s_dqdx, s_dqdy, s_dqdz, s_Aq);
```

**Comment by pdhahn** [comment]
In v0.2, I have to work around this by creating a version of the function which operates with arrays in global address space (i.e., via argument type preceded with occaPointer). Sometimes, this means I create two versions of the same function: e.g., a `gl__myFunc( ... )` version that operates with memory in global address space, and a `myFunc( ... )` version that operates with local address space. This means more cut and paste, or using macros (neither ideal).

**Comment by pdhahn** [comment]
If the v1.x parser could infer not only the address space for/from the calling arguments, but also correspondingly the function code generated/used (possibly multiple versions under the hood with unique signatures), that would be wonderful, since it relieves the problem with cut and paste and/or macros.

**Comment by tcew** [comment]
Paul - We also used macros as work arounds in 0.2 . In 1.0 there have been some issues with the way that the parser interprets macros. I switched a number of macros to device functions and then we ran into the arg type problem of yore. I hope that otherwise OCCA has been useful for you ! Tim

**Comment by pdhahn** [comment]
Love OCCA! It is used extensively in our application development here at my company. The simple C-like programming paradigm is ideal compared to #pragmas or straight-up OpenCL / CUDA embedded within a bunch of #ifdefs. Eagerly waiting for 1.x to become stable!  :-)

BTW following are examples of functions I use for different address spaces (local vs. global). Basically identical function bodies created by cut and paste, differing only by name and use of `occaPointer`.

Yes, this duplication is really only motivated by OpenCL, but that happens to be my most frequently used mode (due to interoperability requirements). But other modes are used of course. And when not in OpenCL mode, such as Serial or OpenMP, that is why the functions have to be separately named (since `occaPointer`, a macro, is defined to be empty for those modes).

Ideally I should only have to write a single function without having to accommodate different address spaces; which is, I believe, a restatement of the gist of this issue.

```
occaFunction
FLOAT_T
getCumulativeSum(
    FLOAT_T*        darr,
    const FLOAT_T*  sarr,
    INT_T           nelms,
    FLOAT_T         xtra_incr    = (FLOAT_T)0.0,
    ///
    /// [in] true means check (and fail) for any
    /// non-positive element in \p sarr
    ///
    int             doCheckMono  = 0
    )
{
    FLOAT_T*  dst = darr;

    const FLOAT_T*  src = sarr;

    const FLOAT_T  pone = (FLOAT_T)1.0;

    FLOAT_T  val;
    FLOAT_T  csum = 0;

    INT_T  n = nelms;

    while( n-- > 0 )
    {
	val = *src++;

	if ( doCheckMono  &&  val < pone )
	    return (FLOAT_T)-1.0;

	csum += val + xtra_incr;

	*dst++ = csum;
    }

    return csum;

}// getCumulativeSum



occaFunction
FLOAT_T
gl__getCumulativeSum(
    occaPointer FLOAT_T*        darr,
    occaPointer const FLOAT_T*  sarr,
    INT_T                       nelms,
    FLOAT_T                     xtra_incr    = (FLOAT_T)0.0,
    ///
    /// [in] true means check (and fail) for any
    /// non-positive element in \p sarr
    ///
    int                         doCheckMono  = 0
    )
{
    occaPointer FLOAT_T*  dst = darr;

    occaPointer const FLOAT_T*  src = sarr;

    const FLOAT_T  pone = (FLOAT_T)1.0;

    FLOAT_T  val;
    FLOAT_T  csum = 0;

    INT_T  n = nelms;

    while( n-- > 0 )
    {
	val = *src++;

	if ( doCheckMono  &&  val < pone )
	    return (FLOAT_T)-1.0;

	csum += val + xtra_incr;

	*dst++ = csum;
    }

    return csum;

}// gl__getCumulativeSum

```

**Comment by pdhahn** [comment]
Please observe how I did a series of edits on both of the above functions, which illustrates the maintenance problem involved with code duplication here.  :-)

**Comment by jeremylt** [comment]
This is showing up on our PR in libCEED : https://github.com/CEED/libCEED/pull/305

**Comment by kris-rowe** [comment]
Closed by #576
